rust/connlib/tunnel/src/tests/sut.rs

```diff
@@ -330,53 +335,14 @@ impl TunnelTest {
     /// Dispatching a [`Transmit`] (read: packet) to a host can trigger more packets, i.e. receiving a STUN request may trigger a STUN response.
     ///
     /// Consequently, this function needs to loop until no host can make progress at which point we consider the [`Transition`] complete.
+    ///
+    /// At most, we will spend 10s of "simulation time" advancing the state.
     fn advance(&mut self, ref_state: &ReferenceState, buffered_transmits: &mut BufferedTransmits) {
-        'outer: loop {
-            if let Some(transmit) = buffered_transmits.pop() {
-                self.dispatch_transmit(transmit, buffered_transmits, &ref_state.global_dns_records);
-                continue;
-            }
-
-            if let Some(transmit) = self.client.exec_mut(|sim| sim.sut.poll_transmit()) {
-                buffered_transmits.push(transmit, &self.client);
-                continue;
-            }
-            if let Some(event) = self.client.exec_mut(|c| c.sut.poll_event()) {
-                self.on_client_event(
-                    self.client.inner().id,
-                    event,
-                    &ref_state.portal,
-                    &ref_state.global_dns_records,
-                );
-                continue;
-            }
-            if let Some(query) = self.client.exec_mut(|client| client.sut.poll_dns_queries()) {
-                self.on_forwarded_dns_query(query, ref_state);
-                continue;
-            }
-            self.client.exec_mut(|sim| {
-                while let Some(packet) = sim.sut.poll_packets() {
-                    sim.on_received_packet(packet)
-                }
-            });
-
-            for (_, gateway) in self.gateways.iter_mut() {
-                let Some(transmit) = gateway.exec_mut(|g| g.sut.poll_transmit()) else {
-                    continue;
-                };
+        let cut_off = self.flux_capacitor.now::<Instant>() + Duration::from_secs(10);
 
-                buffered_transmits.push(transmit, gateway);
-                continue 'outer;
-            }
-
-            for (id, gateway) in self.gateways.iter_mut() {
-                let Some(event) = gateway.exec_mut(|g| g.sut.poll_event()) else {
-                    continue;
-                };
-
-                on_gateway_event(*id, event, &mut self.client, self.now);
-                continue 'outer;
-            }
+        'outer: while self.flux_capacitor.now::<Instant>() < cut_off {
+            self.handle_timeout(&ref_state.global_dns_records, buffered_transmits);
```

Comment:
We always do time-related things first now. This is important because we need to call `handle_timeout` _after_ we have changed other state (like receiving a packet or something). We `continue` in other parts of the loop here meaning if `handle_timeout` is at the top, it always happens after those actions.